keras/engine/training.py

```diff
@@ -999,15 +1005,18 @@ def _fit_loop(self, f, ins, out_labels=None, batch_size=32,
                 print('Train on %d samples, validate on %d samples' %
                       (ins[0].shape[0], val_ins[0].shape[0]))
 
-        if ins and hasattr(ins[0], 'shape'):
-            num_train_samples = ins[0].shape[0]
+        if steps_per_epoch is not None:
+            num_train_samples = steps_per_epoch
```

Comment:
num_train_samples != steps_per_epoch, though. A step is a batch, not a single sample.